include/boost/gil/histogram.hpp

```diff
@@ -0,0 +1,247 @@
+struct has_integral_key_type : std::false_type {} ;
+
+template <typename T>
+struct has_integral_key_type
+        <
+            T,
+            typename std::enable_if<std::is_integral<typename T::key_type >::value >::type 
+        > : std::true_type {} ;
+
+} // namespace detail
+
+/// \ingroup Histogram
+/// \brief Use for indexable and resizable sequence containers 
+///
+/// This overload corresponds to resizable sequence containers which have random access capability.
+//  This would correspond to containers from STL like std::vector and std::deque. The only 
+/// difference from the next function is this function is helpful if older bin values in container
+/// is not to be erased i.e. accumulate the histograms.
+///
+template
+<
+    typename Container,
+    typename SrcView,
+    typename std::enable_if<
+                detail::is_indexable<Container>::value &&
+                detail::is_resizable<Container>::value, int>::type = 0
+>
+inline void image_histogram(SrcView const& srcview, Container &hist)
+{
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    static_assert(std::is_arithmetic<typename Container::value_type>::value,
+                    "Improper container type for images.");
+    static_assert(std::is_unsigned<typename channel_type<SrcView>::type>::value,
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    hist.resize(std::numeric_limits<channel_t>::max() + 1);
+
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        ++hist[p];
+    });
+}
+
+/// \ingroup Histogram
+/// \brief Use for indexable and resizable sequence containers
+///
+/// This overload is the same as the above except that it does not produce the
+/// accumulate effect for histogram bin values. It returns a new container of
+/// size large enough to hold entire range of corresponding pixel value in the input image.
+///
+template
+<
+    typename Container,
+    typename SrcView,
+    typename std::enable_if<
+                detail::is_indexable<Container>::value &&
+                detail::is_resizable<Container>::value, int>::type = 0
+>
+inline Container image_histogram(SrcView const& srcview)
+{
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    static_assert(std::is_arithmetic<typename Container::value_type>::value,
+                    "Improper container type for images.");
+    static_assert(std::is_unsigned<typename channel_type<SrcView>::type>::value,
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    Container hist(std::numeric_limits<channel_t>::max() + 1);
+    
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        ++hist[p];
+    });
```

Comment:
this is code repetition you should consider calling `image_histogram(SrcView const& srcview, Container &hist)`

Reply:
Maybe make a image_histogram_impl(SrcView const& srcview, Container &hist) without the constraints on the template type is what you meant?

Reply:
@lpranam 
> I have some objections with the code format(template parameters)

What are the objections exactly?

Generally, as it was agreed quite some time ago, we try to keep the lines up to [100 characters long](https://github.com/boostorg/gil/blob/develop/CONTRIBUTING.md#guidelines).
Second, parameters are often long (e.g. `mp11::mp_size<typename Layout::color_space_t>::value`) and it is often a good idea to keep names of template parameters self-descriptive, avoid cryptic abbreviation.

So, this for example would fit that limit:

```cpp
template <typename T>
struct is_indexable<T, decltype((void) ++declval<T>()[0] , 0)> : std::true_type {} ;
```

but this would not, hence each parameter goes into separate lines and `<` and `>` are formatted similarly to blocks delimited with `{` and `}` 

```cpp
template <typename ChannelValue, typename Layout>
struct pixel :
    detail::homogeneous_color_base
    <
        ChannelValue,
        Layout,
        mp11::mp_size<typename Layout::color_space_t>::value
    >
{
...
};
```

That makes the whole guideline very simple, without too many exceptions or sub-rules.

The only other, subjectively, readable formatting of highly templated source code that I know is one used by Boost.Spirit or other libraries by [Joel de Guzman](https://github.com/cycfi/), Michael Caisse et. al.

Reply:
@lpranam If you don't have any proposal on better formatting fitting template parameters list in 100 characters long lines, then I'd suggest to mark this conversation as resolved.

Reply:
@mloskot Is it necessary to break 

`histogram<boost::mp11::mp_at<bin_t, boost::mp11::mp_size_t<Dimensions>>...> sub_histogram() {}`

into 

```
    histogram
    <   
        boost::mp11::mp_at
        <
            bin_t,
            boost::mp11::mp_size_t<Dimensions>
        >...
    > sub_histogram() 
    {}
```

The former fits within 100 characters.

Reply:
@codejaeger No, if it fits in line long for 100 characters, then no need for breaking it.
In my earlier comments I just used the snippet based on your code as an example.